123-job-jwt/proposal.md

```diff
@@ -0,0 +1,175 @@
+# Summary
+
+Mint a signed JSON Web Token (JWT) and deliver to the job runtime container.
+Additionally, provide a validation endpoint to allow for external or internal
+validation of the issued JWT.
+
+## Motivation
+
+Bluntly, the native integration with Hashicorp Vault is scoped only to static
+secrets and this implementation is inflexible. By leveraging JWT auth at a job
+level, the claims can become incredibly granular, and allow for flexible policy
```

Comment:
Short answer is yes.

Let's consider a sample JWT:

Header:
```json
{
  "alg": "HS256",
  "typ": "JWT"
}
```
Payload:
```json
{
  "sub": "1234567890",
  "name": "John Doe",
  "team": "main",
  "pipeline": "sample_pipeline"
}
```
...and a [Go example of creating a signature](https://pkg.go.dev/github.com/golang-jwt/jwt#example-New-Hmac) which will be based on the Header and Payload values.

So, as an application development team, we can _literally_ put whatever we want in the header and payload. In the case of Hashicorp Vault, it can only be configured to validate key:string/int claims from the payload, so we want as much of the Concourse runtime metadata in there as possible (Job ID, how job was initiated, user that manually initiated if applicable, team, pipeline, job, step if applicable, timestamps, etc. and **do not use complex structures** like key:map or key:array) so we can configure Vault to look for those things in the payload. (Other systems could leverage the JWT for validation, but I'm keeping things in the Vault context for consistency.)

Then, it all depends on where we're minting and delivering the token. Does each step have a new JWT minted? Or would that be too heavy, and we only mint JWTs per job? The highest level of granularity can be achieved by delivering to the most granular unit, and including that information in the Payload section.

Policy assignment happens on the Vault side, so:
```hcl
# Policy "sample_policy"
path "secret/data/sample_team/sample_pipeline/arbitrary_secret" {
  capabilities = ["read"]
}
```
and an auth method at path `concourse`, plus a role named `main-sample_pipeline`:
```sh
#!/usr/bin/env bash
export VAULT_ADDR=http://127.0.0.1:8200
export VAULT_TOKEN=devenv_root
vault auth enable -path=concourse jwt
vault write auth/concourse/config bound_issuer="concourse-ci.example.internal" jwks_url="https://concourse-ci.example.internal/web/-/jwks" 
vault write -format=json auth/concourse/role/main-sample_pipeline -<<EOF
{
  "bound_claims": {
    "team": "main",
    "pipeline": "sample_pipeline"
  },
  "bound_claims_type": "glob",
  "claim_mappings": {
      "sub": "concourse_subject",
      "name": "concourse_name",
      "team": "concourse_team",
      "pipeline": "concourse_pipeline"
  },
  "role_type": "jwt",
  "token_bound_cidrs": [],
  "token_explicit_max_ttl": 3600,
  "token_max_ttl": 0,
  "token_policies": [
    "sample_policy"
  ],
  "token_ttl": 0,
  "user_claim": "team",
  "clock_skew_leeway": 0, "expiration_leeway": 0, "groups_claim": "", "not_before_leeway": 0, "oidc_scopes": null, "token_no_default_policy": false, "token_num_uses": 0, "token_period": 0, "token_type": "default", "allowed_redirect_uris": null, "bound_audiences": null, "bound_subject": "","verbose_oidc_logging": false
}
```

So, now we:
- Mint a JWT for a job or step
- Send that JWT to the runtime in an Environment Variable
- Step runtime calls Vault at `http://127.0.0.1:8200/v1/auth/concourse/login` with the JSON data payload `{"role": "main-sample_pipeline", "jwt": "xxxxx.yyyyyyy.zzzzzzz"}`
- Vault takes full JWT and sends it to the configured `jwks_url` in the auth method config (`https://concourse-ci.example.internal/web/-/jwks` in the example above)
- Vault gets a response from Concourse saying the JWT is either valid or invalid
- If the JWT is valid, Vault then Base64 decodes the Payload portion of the JWT and compares the claims against the requested role definition.
  - All binds must match:
    - `"bound_claims": {"team": "main","pipeline": "sample_pipeline"},` from role definition
    - `{"sub": "1234567890", "name": "John Doe", "team": "main", "pipeline": "sample_pipeline"}` from JWT Payload
    - Both of the defined `bound_claims` (`team` and `pipeline`) match, so Vault will validate
  - Policies assigned to the role (`sample_policy`) will be given to the resultant token
  - Identity system `entity` claim will be evaluated (usually best not to make this too granular, as in the Enterprise Vault context, each `entity` is a license, so generally all machine auths from the same team can use a single license by making a high level user claim mapping) and policies from the entity and groups will be aggregated (which is usually none or some safe global defaults, because we're saving licenses by mapping the `entity` object to a upper level abstraction layer) when the token is created
- Hashicorp Vault Token is delivered to the Job/Step Runtime, which is used in subsequent calls to Vault where secrets are requested
- The policy only has paths for that pipeline because they logged in using a specific role, and it cannot get secrets that belong to another team
- Vault doesn't have to have a Concourse-specific _secret_ path, it does have to have a Concourse-specific _auth method_ of type `jwt` per Concourse instance though

Further reading:
- https://www.vaultproject.io/docs/auth/jwt
- https://www.vaultproject.io/api-docs/auth/jwt
- https://jwt.io/introduction

Reply:
A note for clarity:

If a Vault JWT auth method role configuration contains:
```json
"bound_claims": {
    "team": ["main", "sample"]
}
```
then a JWT with `{"sub": "1234567890", "name": "John Doe", "team": "main", "pipeline": "sample_pipeline"}` would validate, and so would a JWT with `{"sub": "1234567810", "name": "Jane Doe", "team": "sample", "pipeline": "random_pipeline"}`, but **NOT** a JWT with `{"sub": "1234567890", "name": "John Doe", "team": ["main", "sample"], "pipeline": "sample_pipeline"}`, as Vault does not match claims in this manner. This is why I state that we should avoid putting any complex structures in the JWT payload.